File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys

input = sys.stdin.readline


def dfs(target, row, col, right_up, left_down):
    global answer

    if row == target:
        answer += 1
        return

    for i in range(target):
        leftIdx = i + row
        rightIdx = i - row + target

        if col[i] or right_up[rightIdx] or left_down[leftIdx]:
            continue

        col[i] = True
        right_up[rightIdx] = True
        left_down[leftIdx] = True
        dfs(target, row + 1, col, right_up, left_down)
        col[i] = False
        right_up[rightIdx] = False
        left_down[leftIdx] = False


def solution():
    global answer
    n = int(input())
    answer = 0

    row = 0
    col = [False] * n
    right_up = [False] * (2 * n + 1)
    left_down = [False] * (2 * n + 1)

    dfs(n, row, col, right_up, left_down)
    print(answer)


solution()

File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

    private static int answer = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        int row = 0;
        boolean[] col = new boolean[n];
        boolean[] right_up = new boolean[2 * n + 1];
        boolean[] left_down = new boolean[2 * n + 1];

        dfs(n, row, col, right_up, left_down);
        System.out.println(answer);
    }

    private static void dfs(final int target, final int row, final boolean[] col, final boolean[] rightUp,
                            final boolean[] leftDown) {
        if (row == target) {
            answer++;
            return;
        }

        for (int i = 0; i < target; i++) {
            int leftIdx = i + row;
            int rightIdx = i - row + target;

            if (col[i] || rightUp[rightIdx] || leftDown[leftIdx]) {
                continue;
            }

            col[i] = true;
            rightUp[rightIdx] = true;
            leftDown[leftIdx] = true;
            dfs(target, row + 1, col, rightUp, leftDown);
            col[i] = false;
            rightUp[rightIdx] = false;
            leftDown[leftIdx] = false;
        }
    }
}